File: solutions/4242-sum-of-gcd-of-formed-pairs/sum-of-gcd-of-formed-pairs.go
func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func gcdSum(nums []int) int64 {
	n := len(nums)
	mx := make([]int, n)
	prefixMax := math.MinInt32

	for i, x := range nums {
		if x > prefixMax {
			prefixMax = x
		}
		mx[i] = prefixMax
	}

	prefixGcd := make([]int, n)
	for i := 0; i < n; i++ {
		prefixGcd[i] = gcd(nums[i], mx[i])
	}
	sort.Ints(prefixGcd)

	var ans int64 = 0
	left, right := 0, n-1
	for left < right {
		ans += int64(gcd(prefixGcd[left], prefixGcd[right]))
		left++
		right--
	}

	return ans
}

File: solutions/4242-sum-of-gcd-of-formed-pairs/sum-of-gcd-of-formed-pairs.js
var gcdSum = function (nums) {
    const gcd = (a, b) => {
        while (b !== 0) {
            [a, b] = [b, a % b];
        }
        return a;
    };

    const n = nums.length;
    const mx = new Array(n);
    let prefixMax = -Infinity;
    for (let i = 0; i < n; i++) {
        prefixMax = Math.max(prefixMax, nums[i]);
        mx[i] = prefixMax;
    }

    const prefixGcd = new Array(n);
    for (let i = 0; i < n; i++) {
        prefixGcd[i] = gcd(nums[i], mx[i]);
    }

    prefixGcd.sort((a, b) => a - b);
    let ans = 0;
    let left = 0,
        right = n - 1;
    while (left < right) {
        ans += gcd(prefixGcd[left], prefixGcd[right]);
        left++;
        right--;
    }
    return ans;
};